components/script/dom/servoparser/async_html.rs

```diff
@@ -0,0 +1,561 @@
+        assert!(sink.parse_node_data.borrow_mut().insert(0, data).is_none());
+        sink.nodes.borrow_mut().insert(0, JS::from_ref(document.upcast()));
+        sink
+    }
+
+    fn new_parse_node(&self) -> ParseNode {
+        let id = self.next_parse_node_id.get();
+        let data: ParseNodeData = Default::default();
+        assert!(self.parse_node_data.borrow_mut().insert(id, data).is_none());
+        self.next_parse_node_id.set(id + 1);
+        ParseNode {
+            id: id,
+            qual_name: None,
+        }
+    }
+
+    fn enqueue(&self, op: ParseOperation) {
+        self.process_operation(op);
+    }
+
+    fn process_operation(&self, op: ParseOperation) {
+        let document = Root::from_ref(&**self.nodes.borrow().get(&0).expect("document node should exist!"));
+        let document = document.downcast::<Document>().expect("Document node should be downcasted!");
+        match op {
+            ParseOperation::GetTemplateContents(target, contents) => {
+                let target = Root::from_ref(&**self.nodes.borrow().get(&target).unwrap());
+                let template = target.downcast::<HTMLTemplateElement>().expect(
+                    "Tried to extract contents from non-template element while parsing");
+                assert!(self.nodes.borrow_mut().insert(contents, JS::from_ref(template.Content().upcast())).is_none());
+            }
+            ParseOperation::CreateElement(id, name, attrs) => {
+                let elem = Element::create(name, &*self.document,
+                                           ElementCreator::ParserCreated(self.current_line));
+                for attr in attrs {
+                    elem.set_attribute_from_parser(attr.name, DOMString::from(String::from(attr.value)), None);
+                }
+
+                self.nodes.borrow_mut().insert(id, JS::from_ref(elem.upcast()));
+            }
+            ParseOperation::CreateComment(text, id) => {
+                let comment = Comment::new(DOMString::from(text), document);
+                assert!(self.nodes.borrow_mut().insert(id, JS::from_ref(&comment.upcast())).is_none());
+            }
+            ParseOperation::Insert(parent, reference_child, new_node) => {
+                let nodes = self.nodes.borrow();
+                let parent = nodes.get(&parent).unwrap();
+                let reference_child = reference_child.map(|n| nodes.get(&n).unwrap().deref());
+
+                let node = match new_node {
+                    ParserNodeOrText::Node(ref id) => {
+                        let deref_node = nodes.get(id).unwrap().deref();
+                        NodeOrText::AppendNode(JS::from_ref(deref_node))
+                    },
+                    ParserNodeOrText::Text(ref string) => {
+                        match Tendril::from_str(string) {
+                            Ok(tendril) => NodeOrText::AppendText(tendril),
+                            Err(_) => panic!("Error while converting from string to tendril!"),
```

Comment:
Please file a follow-up issue on Tendril to implement `From<&str>` for `StrTendril`.